tests/v1/test_stats.py

```diff
@@ -0,0 +1,281 @@
+    decoded_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DECODED,
+        monotonic_ts_s=decoded_ts,
+        num_new_tokens=1,
+        token_perf_ts_ns=decoded_ts * 1e9,
+    )
+    stats.update_from(decoded_update)
+    assert stats.last_updated_ts_s == decoded_ts
+    # Since arrival
+    assert stats.first_token_latency_s == decoded_ts - arrived_ts
+    assert stats.num_output_tokens == 1
+    # Since first scheduled
+    assert stats.prefill_latency_s == 2
+
+    # Test DETOKENIZED
+    detokenized_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DETOKENIZED,
+        monotonic_ts_s=detokenized_ts,
+    )
+    stats.update_from(detokenized_update)
+    assert stats.last_updated_ts_s == detokenized_ts
+
+    # Test another DECODE should yield correct inter token latency
+    decoded_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DECODED,
+        monotonic_ts_s=decoded_2_ts,
+        num_new_tokens=1,
+        token_perf_ts_ns=decoded_2_ts * 1e9,
+    )
+    stats.update_from(decoded_update)
+    assert stats.output_token_latency_s_lst == [
+        decoded_2_ts - decoded_ts,
+    ]
+    assert stats.num_output_tokens == 2
+
+    # Test PREEMPTED
+    preempted_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.PREEMPTED,
+        monotonic_ts_s=preempted_ts,
+    )
+    stats.update_from(preempted_update)
+    assert stats.last_updated_ts_s == preempted_ts
+    assert stats.preempted_ts_s_lst == [preempted_ts]
+    # States should be reset
+    assert stats.num_computed_tokens == 0
+    assert stats.num_cached_tokens == 0
+    assert stats.num_output_tokens == 0
+    assert stats.output_token_latency_s_lst == []
+
+    # Test resumed
+    resumed_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.RUNNING,
+        monotonic_ts_s=resumed_ts,
+        new_prefill=True,
+        num_computed_tokens=6,
+        num_cached_tokens=2,
+    )
+    stats.update_from(resumed_update)
+    # Resumed prefill timestamp should be updated
+    assert stats.prefill_ts_s == resumed_ts
+    assert stats.num_computed_tokens == 6
+    assert stats.num_cached_tokens == 2
+    assert stats.prefill_start_ts_s_lst == [
+        running_ts,
+        resumed_ts,
+    ]
+    assert stats.last_updated_ts_s == resumed_ts
+
+    # Test another DECODED should yield correct first token latency.
+    decoded_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DECODED,
+        monotonic_ts_s=decoded_3_ts,
+        num_new_tokens=1,
+        token_perf_ts_ns=decoded_3_ts * 1e9,
+    )
+    stats.update_from(decoded_update)
+    assert stats.first_token_ts_s == decoded_3_ts
+    assert stats.num_output_tokens == 1
+
+    # Test FINISHED
+    finished_update = RequestStatsUpdate(
```

Comment:
Is `FINISHED` going to be from the POV of the `EngineCore` or the `Detokenizer`?

I could see an argument for either.

One thing to note, this has a nasty intersection with how stop strings are handled. The way it works is that we detect the stop string in the `detokenizer`, we finish the request in the `detokenizer` and send an `abort` message to the `EngineCore`

Reply:
So yeah, FINISH will currently be from either detokenizer or enginecore but not both, depending on the finish reasons:

1. From detokenizer: it's essentially due to stop string --> the finish from engine core will not override the detokenizer's finish reason. 
2. Else: enginecore decides to finish a request --> the detokenizer's finish scenario won't kick in.

Reply:
I think maybe we could unify the two by always recording a request is finished from the engine core's perspective. 

When detokenizer finishes a request due to stop string, we could include the finish reason in the abort request RPC,  then we will be able to have a single point (from engine core) when a request is finished for all reasons (at least those I think we support now?) 
- Finished due to engine core limit and sampling param (stop token, max len, and etc) 
- Stop string
- Users manually abort